File: mobile/android/FUN/app/src/main/java/com/fun/app/data/models/PlaylistContext.kt
/**
 * FUN App - Playlist Context Models
 */

package com.fun.app.data.models

import com.squareup.moshi.JsonClass

enum class PlaylistMode {
    DISCOVER,
    BINGE,
    SERIES
}

data class PlaylistContext(
    val mode: PlaylistMode,
    val episodes: List<Episode>,
    var currentIndex: Int,
    val seriesId: String?,
    val seriesTitle: String?
) {
    val currentEpisode: Episode?
        get() = episodes.getOrNull(currentIndex)
    
    val nextEpisode: Episode?
        get() = episodes.getOrNull(currentIndex + 1)
    
    val previousEpisode: Episode?
        get() = episodes.getOrNull(currentIndex - 1)
    
    val hasNext: Boolean
        get() = currentIndex + 1 < episodes.size
    
    val hasPrevious: Boolean
        get() = currentIndex > 0
    
    val position: Pair<Int, Int>
        get() = Pair(currentIndex + 1, episodes.size)
    
    fun moveNext(): Boolean {
        return if (hasNext) {
            currentIndex++
            true
        } else {
            false
        }
    }
    
    fun movePrevious(): Boolean {
        return if (hasPrevious) {
            currentIndex--
            true
        } else {
            false
        }
    }
}

File: mobile/ios/FUN/FUN/Models/PlaylistContext.swift
/**
 * FUN App - Playlist Context Models
 */

import Foundation

enum PlaylistMode: String, Codable {
    case discover = "discover"
    case binge = "binge"
    case series = "series"
}

struct PlaylistContext {
    let mode: PlaylistMode
    var episodes: [Episode]
    var currentIndex: Int
    let seriesId: String?
    let seriesTitle: String?
    
    var currentEpisode: Episode? {
        guard currentIndex >= 0 && currentIndex < episodes.count else { return nil }
        return episodes[currentIndex]
    }
    
    var nextEpisode: Episode? {
        guard currentIndex + 1 < episodes.count else { return nil }
        return episodes[currentIndex + 1]
    }
    
    var previousEpisode: Episode? {
        guard currentIndex > 0 else { return nil }
        return episodes[currentIndex - 1]
    }
    
    var hasNext: Bool {
        return currentIndex + 1 < episodes.count
    }
    
    var hasPrevious: Bool {
        return currentIndex > 0
    }
    
    var position: (current: Int, total: Int) {
        return (currentIndex + 1, episodes.count)
    }
    
    mutating func moveNext() -> Bool {
        guard hasNext else { return false }
        currentIndex += 1
        return true
    }
    
    mutating func movePrevious() -> Bool {
        guard hasPrevious else { return false }
        currentIndex -= 1
        return true
    }
}
